### PR TITLE
Issue #11014 - Cleanup of relative redirect handling Jetty-12

### LINKED ISSUE #11014: RedirectRegexRule and RewritePatternRule should consider relativeRedirectAllowed
There is an HttpConfiguration option "relativeRedirectAllowed" that is taken into consideration by sendRedirect().  But RedirectRegexRule and RewritePatternRule don't use that sendRedirect.  Perhaps they should?

### PR DESCRIPTION
**Pull request by gregw**
+ Handle request relative redirects
+ Moved to Response
+ Changed default to allow relative

Fixes #11014

### CODE REVIEW

**Comment by gregw** [comment]
@sbordet @joakime @lorban nudge!

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpConfiguration.java [hunk 1]
```diff
@@ -76,7 +76,7 @@ public class HttpConfiguration implements Dumpable
     private CookieCompliance _requestCookieCompliance = CookieCompliance.RFC6265;
     private CookieCompliance _responseCookieCompliance = CookieCompliance.RFC6265;
     private boolean _notifyRemoteAsyncErrors = true;
-    private boolean _relativeRedirectAllowed;
+    private boolean _relativeRedirectAllowed = true;
```
Comment (joakime) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpConfiguration.java:79]: (for other reviewers) This changes the default to always use relative redirects (path only).
Prior versions of Jetty would try to always use absolute URIs.

**Review by joakime — DISMISSED** [review]

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java [hunk 2]
```diff
@@ -604,45 +604,18 @@ static List<HttpCookie> getCookies(Request request)
...
-            if (location == null)
-                throw new IllegalStateException("redirect path cannot be above root");
-            url.append(location);
-
-            location = url.toString();
-        }
-        // TODO do we need to do request relative without scheme?
-
-        return location;
+        return Response.toRedirectURI(request, location);
```
Comment (sbordet) [inline jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:618]: I would have left the original code, and not moved it to `Response`.
The response is never used in the implementation, and enforces that in order to build a redirect URI, you need a `Request` (not a `Response`).

File: jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ResponseTest.java [hunk 3]
```diff
@@ -233,9 +235,67 @@ public boolean handle(Request request, Response response, Callback callback)
...
+
+        request = """
+                GET /path/to/ HTTP/1.1\r
+                Host: hostname\r
+                Connection: close\r
+                \r
+                """;
+        response = HttpTester.parseResponse(connector.getResponse(request));
+        assertEquals(HttpStatus.MOVED_TEMPORARILY_302, response.getStatus());
+        assertThat(response.get(HttpHeader.LOCATION), is("/path/to/somewhere/else"));
```
Comment (sbordet) [inline jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/ResponseTest.java:317]: Can we have a test that clarifies whether the location must be already encoded or not, especially for query values?

**Review by sbordet — CHANGES_REQUESTED** [review]

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java (hunk 2 above)
Comment (gregw) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:618]: @sbordet, Sending a redirect is a response function and it naturally lives next to the sendRedirect methods.  Plus we often pass the request and the response into these static methods e.g. `Response.sendRedirect(Request, Response, Callback, location)`

File: jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpConfiguration.java (hunk 1 above)
Comment (lorban) [reply jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/HttpConfiguration.java:79]: I don't mind the default behavior change, but maybe it should be justified with a comment?

**Review by lorban — APPROVED** [review]

**Review by joakime — APPROVED** [review]

**Review by sbordet — APPROVED** [review]

**Pull request merged**